README.md

```diff
@@ -2,302 +2,708 @@
-class OrdersSearch
-  include Parametric::Params
-  param :query, 'Search query. optional', nullable: true
-  param :tags, 'Tags', multiple: true
+required_name_schema = Parametric::Schema.new do
+  field(:name).required
+  field(:age)
 end
 
-search = OrdersSearch.new({})
-search.params # {tags: []}
+optional_name_schema = Parametric::Schema.new do
+  field(:name)
+end
+
+# This schema now has :name and :age fields.
+# :name has been redefined to not be required.
+new_schema = required_name_schema.merge(optional_name_schema)
 ```
 
-## `available_params`
+## #meta
 
-`#available_params` returns the subset of keys that were populated (including defaults). Useful for building query strings.
+The `#meta` field method can be used to add custom meta data to field definitions.
+These meta data can be used later when instrospecting schemas (ie. to generate documentation or error notices).
 
 ```ruby
-order_search = OrdersSearch.new(page: 2, foo: 'bar')
-order_search.available_params # => {page: 2, per_page: 50}
+create_user_schema = Parametric::Schema.do
+  field(:name).required.type(:string).meta(label: "User's full name")
+  field(:status).options(["published", "unpublished"]).default("published")
+  field(:age).type(:integer).meta(label: "User's age")
+  field(:friends).type(:array).meta(label: "User friends").schema do
+    field(:name).type(:string).present.meta(label: "Friend full name")
+    field(:email).validate(:email).meta(label: "Friend's email")
+  end
+end
 ```
 
-## `schema`
+## #schema
 
-`#schema` returns a data structure including meta-data on each parameter, such as "label" and "options". Useful for building forms or self-documented Hypermedia APIs (or maybe [json-schema](http://json-schema.org/example2.html) endpoints).
+A `Schema` instance has a `#schema` method that allows instrospecting schema meta data.
 
 ```ruby
-order_search.schema[:q].label # => 'Full text search query'
-order_search.schema[:q].value # => ''
+create_user_schema.schema[:name].label # => "User's full name"
+create_user_schema.schema[:age].label # => "User's age"
+create_user_schema.schema[:friends].label # => "User friends"
+# Recursive schema data
+create_user_schema.schema[:friends].schema[:name].label # => "Friend full name"
+```
 
-order_search.schema[:page].label # => 'Page number'
-order_search.schema[:page].value # => 1
+Note that many field methods add field meta data.
 
-order_search.schema[:status].label # => 'Order status'
-order_search.schema[:status].value # => ['pending']
-order_search.schema[:status].options # => ['checkout', 'pending', 'closed', 'shipped']
-order_search.schema[:status].multiple # => true
-order_search.schema[:status].default # => 'closed'
+```ruby
+create_user_schema.schema[:name].type # => :string
+create_user_schema.schema[:name].required # => true
+create_user_schema.schema[:status].options # => ["published", "unpublished"]
+create_user_schema.schema[:status].default # => "published"
 ```
 
-## Coercing values
+## #walk
```

Comment:
It is indeed a visitor. It uses an underlying method call `#visit`, in fact. I decided not to call it `#visit` because I don't think the visitor patterns is expected to _collect_ the ouput as `map` does. `map` would be confusing since this version is recursive. `reduce` would probably be misleading...